src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+                  "provided Error Visitor is not exhaustive");
+    // Throwing an exception isn't the sole way to signalize an error
+    // with it. This approach nicely fits cold, infrequent issues but
+    // when applied to a hot one (like ENOENT on write path), it will
+    // likely hurt performance.
+    // Alternative approach for hot errors is to create exception_ptr
+    // on our own and place it in the future via make_exception_future.
+    // When ::handle_exception is called, handler would inspect stored
+    // exception whether it's hot-or-cold before rethrowing it.
+    // The main advantage is both types flow through very similar path
+    // based on future::handle_exception.
+    //
+    // NOTE: this is extension of language. It should be available both
+    // in GCC and Clang but a fallback (basing on throw-catch) could be
+    // added as well.
+    if (type_info == ErrorT::get_exception_ptr_type_info()) {
+      // set `state::invalid` in internals of `seastar::future` to not
+      // call `report_failed_future()` during `operator=()`.
+      [[maybe_unused]] auto&& ep = std::move(result).get_exception();
+
+      using ReturnType = std::result_of_t<ErrorVisitorT(ErrorT)>;
+      if constexpr (std::is_assignable_v<decltype(result), ReturnType>) {
+        result = std::invoke(std::forward<ErrorVisitorT>(errfunc),
+                             ErrorT::from_exception_ptr(std::move(ep)));
+      } else if constexpr (!std::is_same_v<ReturnType, void>) {
+        // void denotes explicit discarding
+        static_assert(_impl::always_false<ReturnType>::value,
+                      "return of Error Visitor is not assignable to future");
+        // do nothing with `ep`.
+      }
+    }
+  }
+
+  auto get_result() && {
+    return std::move(result);
+  }
+};
+
+template <class FuncHead, class... FuncTail>
+static constexpr auto composer(FuncHead&& head, FuncTail&&... tail) {
+  return [
+    head = std::forward<FuncHead>(head),
+    // perfect forwarding in lambda's closure isn't availble in C++17
+    // using tuple as workaround; see: https://stackoverflow.com/a/49902823
+    tail = std::make_tuple(std::forward<FuncTail>(tail)...)
+  ] (auto&&... args) mutable -> decltype(auto) {
+    if constexpr (std::is_invocable_v<FuncHead, decltype(args)...>) {
+      return std::invoke(std::forward<FuncHead>(head),
+                         std::forward<decltype(args)>(args)...);
+    } else if constexpr (sizeof...(FuncTail) > 0) {
+      using next_composer_t = decltype(composer<FuncTail...>);
+      auto&& next = std::apply<next_composer_t>(composer<FuncTail...>,
+                                                std::move(tail));
+      return std::invoke(std::move(next),
+                         std::forward<decltype(args)>(args)...);
+    } else {
+      static_assert(
+        _impl::always_false<FuncHead>::value, "composition is not exhaustive");
+    }
+  };
+}
+
+template <class... WrappedAllowedErrorsT>
+struct errorator {
```

Comment:
Quibble: should this be in crimson/common or something?  The specific errors are specific to osd/os, but we'll probably use this framework all over the place?

Reply:
Yeah, it should be moved. `crimson/common` seems to be good fit.

Reply:
Probably worth doing sooner rather than later to avoid needing to update a bunch of includes.